### PR TITLE
clang: Add pragma clang fp reciprocal

### LINKED ISSUE #64798: clang pragma fp should allow emission of IR arcp flag
pragma clang fp already exists which enables fast math flags within a scope. It currently supports `pragma clang fp contract(value)` and `pragma clang fp reassociate(on)`. 

I would like this extended to cover other fast math flags, in particular arcp. For something like

```
float fast_div(float x, float y) {
  #pragma clang fp reciprocal(on)
  return x / y;
}
```

I would hope to get
```
float @fast_div(float %, float %y) {
   %0 = fdiv arcp float %x, %y
   ret float %0 
}

```

### PR DESCRIPTION
**Pull request by arsenm**
Just follow along with the reassociate pragma. This allows locally setting the arcp fast math flag. Previously you could only access this through the global -freciprocal-math.

Fixes #64798

### CODE REVIEW

**Review by andykaylor — COMMENTED** [review]
Thanks for adding this! I have just a few minor comments.

File: clang/include/clang/Basic/DiagnosticParseKinds.td
[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/include/clang/Basic/DiagnosticParseKinds.td:1]: and a test for it.

**Comment by zahiraam** [comment]
Can you please fix the description of the patch.
Please note that Melanie has retired. You can add me as a reviewer for all FP related work. Thanks.

File: clang/lib/Sema/SemaAttr.cpp [hunk 1]
```diff
@@ -1309,6 +1309,13 @@ void Sema::ActOnPragmaFPReassociate(SourceLocation Loc, bool IsEnabled) {
   CurFPFeatures = NewFPFeatures.applyOverrides(getLangOpts());
 }
 
+void Sema::ActOnPragmaFPReciprocal(SourceLocation Loc, bool IsEnabled) {
```
[the remark this answers is not in the document]
Comment (arsenm) [reply clang/lib/Sema/SemaAttr.cpp:1312]: Why does the fp eval method matter here?

[the remark this answers is not in the document]
Comment (andykaylor) [reply clang/lib/Sema/SemaAttr.cpp:1312]: There's some discussion of it here: https://reviews.llvm.org/D122155

The basic idea is that it doesn't make sense to allow value-changing optimizations while also specifying extra precision to be used for intermediate calculations.

File: clang/include/clang/Basic/DiagnosticParseKinds.td
[the remark this answers is not in the document]
Comment (arsenm) [reply clang/include/clang/Basic/DiagnosticParseKinds.td:1]: Added the string to the message. It's covered by the existing test

**Comment by arsenm** [comment]
> Can you please fix the description of the patch. 

Fix what about it?

**Comment by zahiraam** [comment]
> Just follow allow with the reassociate pragma.

What does that mean?

File: clang/test/CodeGen/fp-reciprocal-pragma.cpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+// Pragma does not propagate through called functions
+float helper_func(float a, float b, float c) { return a + b + c; }
+float fp_recip_call_helper(float a, float b, float c) {
+// CHECK-LABEL: _Z20fp_recip_call_helperfff
+// CHECK: %[[S1:.+]] = fadd float %a, %b
+// CHECK-NEXT: fadd float %[[S1]], %c
+#pragma clang fp reciprocal(on)
+  return helper_func(a, b, c);
+}
```
Comment (yxsamliu) [inline clang/test/CodeGen/fp-reciprocal-pragma.cpp:130]: need a test for interaction between pragma for reassociate and recriprocal, e.g. both on

Comment (arsenm) [reply clang/test/CodeGen/fp-reciprocal-pragma.cpp:130]: That's already here on line 44

**Comment by arsenm** [comment]
ping

**Review by yxsamliu — APPROVED** [review]
LGTM. Thanks

**Comment by zahiraam** [comment]
LGTM.

**Review by zahiraam — APPROVED** [review]

**Pull request merged**